src/common/application/song-database-factory.ts

```diff
@@ -0,0 +1,135 @@
+import {MaiToolsApi} from '../infra/mai-tools-api';
+import {MagicApi} from '../infra/magic-api';
+import {GameVersion, LATEST_VERSION} from '../game-version';
+import {GameRegion} from '../game-region';
+import {SongDatabase, SongProperties} from '../song-props';
+import {getRemovedSongs} from '../removed-songs';
+import {ChartType} from '../chart-type';
+
+export class SongDatabaseFactory {
+  constructor(
+    private readonly maiToolsApi: MaiToolsApi,
+    private readonly magicApi: MagicApi,
+  ) {
+  }
+
+  async create(
+    gameVer: GameVersion,
+    gameRegion: GameRegion,
+  ): Promise<SongDatabase> {
+    const dxMap = new Map<string, SongProperties>();
+    const standardMap = new Map<string, SongProperties>();
+    const nameByIco = new Map<string, string>();
+
+    const songs = await this.magicApi.loadMagic(gameVer);
+    for (const song of songs) {
+      this.insertOrUpdateSong(dxMap, standardMap, nameByIco, song);
+    }
+
+    const chartLevelOverrides = await this.maiToolsApi.fetchChartLevelOverrides(gameVer);
+    // console.log('chartLevelOverrides', chartLevelOverrides);
+    for (const songProps of chartLevelOverrides) {
+      this.insertOrUpdateSong(dxMap, standardMap, nameByIco, songProps);
+    }
+
+    const regionOverrides = await this.maiToolsApi.fetchRegionOverrides(gameRegion);
+    // console.log('regionOverrides', regionOverrides);
+    for (const songProps of regionOverrides) {
+      this.updateSong(dxMap, standardMap, nameByIco, songProps);
+    }
+
+    const removedSongs = getRemovedSongs(gameRegion);
+    for (const songName of removedSongs) {
+      this.deleteSong(dxMap, standardMap, songName);
+    }
+
+    this.validate(dxMap, standardMap);
+    return new SongDatabase(
+      gameVer,
+      gameRegion,
+      true,
+      dxMap,
+      standardMap,
+      nameByIco,
+    );
+  }
+
+  private insertOrUpdateSong(
+    dxMap: Map<string, SongProperties>,
+    standardMap: Map<string, SongProperties>,
+    nameByIco: Map<string, string>,
+    song: SongProperties,
+  ) {
+    const map = song.dx === ChartType.DX ? dxMap : standardMap;
+    if (this.updateSong(dxMap, standardMap, nameByIco, song)) {
+      return;
+    }
+    const key = song.nickname ?? song.name;
```

Comment:
There is a reason why I only use nickname when the song name is Link. The magic contains many nicknames that I can't derive from maimai NET page. If nicknames are used, those songs won't be able to find the SongProperties.

You can try using score-sort on https://maimaidx-eng.com/maimai-mobile/record/musicGenre/search/?genre=101&diff=3 and check あの世行きのバスに乗ってさらば。 or 真・ハンサム体操でズンドコホイ. They probably would show as 13.0~ instead of their actual chart constants